paragraphs.type_clone.inc

```diff
@@ -0,0 +1,423 @@
+function paragraphs_type_clone_get_next_type_number($paragraph_type_prefix) {
+  // Get paragraph types.
+  $paragraph_types = array_keys(paragraphs_bundle_load());
+
+  // Prepare the search pattern.
+  $pattern = '/^' . $paragraph_type_prefix . '/';
+
+  // Count existing prefix occurrences.
+  $occurrences = count(preg_grep($pattern, $paragraph_types));
+
+  // Return an incremented value.
+  return $occurrences + 1;
+}
+
+/**
+ * Form validation handler for the clone creation form.
+ *
+ * @param array $form
+ *   The form array.
+ * @param array $form_state
+ *   Reference to the form state array.
+ */
+function paragraphs_admin_bundle_clone_create_form_validate(array $form, array &$form_state) {
+  if ($form['#form_id'] == 'paragraphs_admin_bundle_clone_create_form') {
+    // Get paragraph types.
+    $paragraph_types = array_keys(paragraphs_bundle_load());
+
+    // If the machine name already exists.
+    if (in_array($form_state['values']['target_machine_name'], $paragraph_types)) {
+      form_set_error('target_machine_name', t('The paragraph type machine name provided already exists'));
+    }
+
+    // Add the custom form submit handler.
+    $form['#submit'][] = 'paragraphs_admin_bundle_clone_create_form_submit';
+  }
+}
+
+/**
+ * Form submit handler for the clone creation form.
+ *
+ * @param array $form
+ *   The form array.
+ * @param array $form_state
+ *   Reference to the form state array.
+ */
+function paragraphs_admin_bundle_clone_create_form_submit(array $form, array &$form_state) {
+  if ($form['#form_id'] == 'paragraphs_admin_bundle_clone_create_form') {
+    // Execute the function batch function.
+    $batch = paragraphs_type_clone_create($form_state['values']);
+    batch_set($batch);
+  }
+}
+
+/**
+ * Create a paragraph type clone.
+ *
+ * This function creates the new paragraph type and adds the cloned fields.
+ *
+ * @param array $values
+ *   Form values coming from the paragraph type creation form.
+ */
+function paragraphs_type_clone_create(array $values) {
+
+  // Create the paragraph type object and populate from the form.
+  $bundle = new stdClass();
+  $bundle->bundle = $values['target_machine_name'];
+  $bundle->locked = 1;
+  $bundle->name = trim((string) $values['target_name']);
+  $bundle->allow_unpublish = (int) $values['target_allow_unpublish'];
+  $bundle->label = (empty($values['target_label'])) ? $bundle->name : trim($values['target_label']);
+  $bundle->description = trim((string) $values['target_description']);
+
+  // Save the paragraph type.
+  paragraphs_bundle_save($bundle);
+
+  // Prepare the operations array.
+  $operations = array();
+
+  // Get paragraph type base fields.
+  $base_fields = field_info_instances('paragraphs_item', $values['source_machine_name']);
+
+  // Get paragraph type extra fields.
+  $extra_fields = field_info_extra_fields('paragraphs_item', $values['source_machine_name'], 'form');
+
+  // Merge all fields.
+  $all_fields = array_merge($base_fields, $extra_fields);
```

Comment:
@yorkshire-pudding If we check first to see if `$base_fields` is empty, can we skip this whole section below and avoid the need for the extra code and checks later in the process?